src/Agent.Listener/Configuration/ConfigurationProvider.cs

```diff
@@ -47,53 +49,70 @@ public override void Initialize(IHostContext hostContext)
             _agentServer = HostContext.GetService<IAgentServer>();
         }
 
-        public void UpdateAgentSetting(AgentSettings settings)
-        {
-            // No implementation required
-        }
-
-        public string GetServerUrl(CommandSettings command)
+        public void GetServerUrl(AgentSettings agentSettings, CommandSettings command)
         {
-            return command.GetUrl();
+            agentSettings.ServerUrl =  command.GetUrl();
         }
 
-        public async Task<int> GetPoolId(CommandSettings command)
+        public async Task GetPoolId(AgentSettings agentSettings, CommandSettings command)
         {
             int poolId = 0;
             string poolName;
 
             poolName = command.GetPool();
             poolId = await GetPoolIdAsync(poolName);
             Trace.Info($"PoolId for agent pool '{poolName}' is '{poolId}'.");
- 
-            return poolId;
+
+            agentSettings.PoolId = poolId;
         }
 
         public string GetFailedToFindPoolErrorString() => StringUtil.Loc("FailedToFindPool");
 
-        public Task<TaskAgent> UpdateAgentAsync(int poolId, TaskAgent agent, CommandSettings command)
+        public void ThrowTaskAgentExistException(AgentSettings agentSettings)
         {
-            return _agentServer.UpdateAgentAsync(poolId, agent);
+            throw new TaskAgentExistsException(StringUtil.Loc("AgentWithSameNameAlreadyExistInPool", agentSettings.PoolId, agentSettings.AgentName));
         }
 
-        public Task<TaskAgent> AddAgentAsync(int poolId, TaskAgent agent, CommandSettings command)
+        public Task<TaskAgent> UpdateAgentAsync(AgentSettings agentSettings, TaskAgent agent, CommandSettings command)
+        {
+            return _agentServer.UpdateAgentAsync(agentSettings.PoolId, agent);
+        }
+
+        public Task<TaskAgent> AddAgentAsync(AgentSettings agentSettings, TaskAgent agent, CommandSettings command)
         { 
-            return _agentServer.AddAgentAsync(poolId, agent);
+            return _agentServer.AddAgentAsync(agentSettings.PoolId, agent);
         }
 
-        public Task DeleteAgentAsync(int agentPoolId, int agentId)
+        public async Task DeleteAgentAsync(AgentSettings agentSettings)
         {
-            return _agentServer.DeleteAgentAsync(agentPoolId, agentId);
+            string currentAction = StringUtil.Loc("UnregisteringAgent");
```

Comment:
but in case agent is not available with server we need to add log like - skipping agent removal